### PR TITLE
HIVE-25898: Compaction txn heartbeating after Worker timeout

### PR DESCRIPTION
**Pull request by veghlaci05**
### What changes were proposed in this pull request?

This PR fixes the issue when the Heartbeater of the CompactionTxn is not shut down. This could happen if an exception is thrown from CompactionTxn.abort() or CompactionTxn.commit() which is not caught by the catch block. The fix reworks heartbeat handling the following way:
1. The heartbeater is shut down BEFORE calling abort() or commit(), so regardless of the outcome of the transaction, or if an exception is thrown or not, the heartbeater won't stay alive. Another benefit that, this approach solves the logging after commit()/abort() issue too, without the need of the previously introduced additional logic (See: HIVE-25740).

### Why are the changes needed?
To fix the issue that in some cases the heartbeater of a CompactionTxn is not shut down properly.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?

Through unit tests.

### CODE REVIEW

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 1]
```diff
@@ -170,6 +176,82 @@ public void tearDown() {
...
+    rqst.setPartitionname("bkt=0");
+    txnHandler.compact(rqst);
+
+    Worker worker = Mockito.spy(new Worker());
+    worker.setThreadId((int) worker.getId());
+    worker.setConf(conf);
+    worker.init(new AtomicBoolean(true));
+    FieldSetter.setField(worker, RemoteCompactorThread.class.getDeclaredField("msc"), mockedClient);
+
+    class Capture<T> implements Answer {
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:233]: i think it's more readable:
````
    AtomicReference<Worker.CompactionTxn> capture = new AtomicReference<>();
    
    doAnswer(invocation -> {
      Worker.CompactionTxn compactionTxn = (Worker.CompactionTxn) invocation.callRealMethod();
      capture.set(compactionTxn);
      return compactionTxn;
    }).when(worker).newCompactionTxn();

    worker.run();

    //Check if the heartbeating is properly terminated
    Assert.assertTrue(capture.get().isHeartbeatTerminated());
````

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 2]
```diff
@@ -170,6 +176,82 @@ public void tearDown() {
...
+            " values(57, 'Budapest')", driver);
+    executeStatementOnDriver("INSERT INTO TABLE " + tblName + " PARTITION(bkt=1)" +
+            " values(58, 'Milano')", driver);
+    execSelectAndDumpData("select * from " + tblName, driver, "Dumping data for " +
+            tblName + " after load:");
+
+    TxnStore txnHandler = TxnUtils.getTxnStore(conf);
+
+    // Commit will throw an exception
+    final IMetaStoreClient mockedClient = Mockito.spy(HiveMetaStoreUtils.getHiveMetastoreClient(conf));
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:219]: use 
````
final IMetaStoreClient mockedClient = Mockito.spy(new HiveMetaStoreClient(conf));
````
and remove MockMaker

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 3]
```diff
@@ -749,44 +740,39 @@ long getLockId() {
...
-      }
+    private void commit() throws TException {
       if (status == TxnStatus.OPEN) {
-        try {
-          msc.commitTxn(txnId);
-          status = TxnStatus.COMMITTED;
-        } catch (TException e) {
-          LOG.error("Caught an exception while committing compaction txn in worker " + workerName, e);
-        }
+        msc.commitTxn(txnId);
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:756]: shouldn't we restore null checks in this PR?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 4]
```diff
@@ -715,28 +705,29 @@ void wasSuccessful() {
...
         }
       }
-      shutdownHeartbeater();
     }
 
     private void shutdownHeartbeater() {
       if (heartbeatExecutor != null) {
-        heartbeatExecutor.shutdownNow();
         try {
+          heartbeatExecutor.shutdownNow();
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:722]: let's move it out of the try-catch

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 5]
```diff
@@ -696,11 +682,15 @@ void open(CompactionInfo ci) throws TException {
             + "}, status {" + res.getState() + "}, reason {" + res.getErrorMessage() + "}");
       }
       lockId = res.getLockid();
-
-      heartbeatExecutor = Executors.newSingleThreadScheduledExecutor();
+      heartbeatExecutor = Executors.newSingleThreadScheduledExecutor(
+              CompactorUtil.createThreadFactory("CompactionTxn Heartbeater %d"));
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:686]: what would be the id, would it be compactionTxn.getTxnId() ?

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java (hunk 2 above)
Comment (veghlaci05) [reply itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:219]: I can replace the factory call with the constructor, but without the mockmaker, the capture does not work at all. It seems capturing method results also require this feature to be enabled.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 6]
```diff
@@ -339,15 +326,19 @@ protected Boolean findNextCompactionAndExecute(boolean collectGenericStats, bool
...
-        }
+
+    // If an exception is thrown in the try-with-resources block below, msc is closed and nulled, so a new instance
+    // is need to be obtained here.
+    if (msc == null) {
+      try {
+        msc = HiveMetaStoreUtils.getHiveMetastoreClient(conf);
+      } catch (Exception e) {
+        LOG.error("Failed to connect to HMS", e);
+        return false;
```
Comment (deniskuzZ) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:337]: shouldn't we throw an exception here?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 5 above)
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:686]: No, that is a builtin feature, it will be a sequential number assigned by the thread factory

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 3 above)
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:756]: Since the reinit of msc field moved out of the try with resources block, it is no longer necessary.

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 7]
```diff
@@ -22,6 +22,7 @@
 import static org.apache.hadoop.hive.ql.TestTxnCommands2.runInitiator;
 import static org.apache.hadoop.hive.ql.TestTxnCommands2.runWorker;
 import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.*;
```
Comment (deniskuzZ) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:25]: why wildcard import?

**Review by deniskuzZ — APPROVED** [review]
pending tests

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 3 above)
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:753]: I see places where we are explicitly setting msc to null. I don't know well this part of the code but if null checks are redundant wouldn't be better to treat it in a separate PR?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 8]
```diff
@@ -715,28 +705,26 @@ void wasSuccessful() {
...
         heartbeatExecutor.shutdownNow();
         try {
           if (!heartbeatExecutor.awaitTermination(5, TimeUnit.SECONDS)) {
-            heartbeatExecutor.shutdownNow();
+            LOG.debug("Heartbeating for transaction {} did not stop in 5 seconds, do not wait any longer.", this);
           }
-          LOG.debug("Successfully stopped heartbeating for transaction {}", this);
         } catch (InterruptedException ex) {
-          heartbeatExecutor.shutdownNow();
+          //Caller thread was interrupted while waiting for heartbeater to terminate, nothing to do
```
[the remark this answers is not in the document]
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:718]: I don't think it is necessary, that log would be sth like "Compaction was aborted while waiting for the heartbeater to shut down" which doesn't add too much in my opinion

[the remark this answers is not in the document]
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:718]: If there is an interrupt during `heartbeatExecutor#awaitTermination` we will never see another log (nor stacktrace) that there was an attempt to shutdown the heartbeater. Moreover, I don't feel very comfortable about swallowing completely an interrupted exception. The least that I would expect here is:

`Thread.currentThread().interrupt();`

but if you have thoroughly though about it and it is not necessary then I trust your judgement.

TLDR 
Citing the "Java Concurrency in Practice":
Propagate the InterruptedException . This is often the most sensible policy if you can get away with it just propagate the InterruptedException to your caller. This could involve not catching InterruptedException , or catching it and throwing it again after performing some brief activity specific cleanup.

Restore the interrupt. Sometimes you cannot throw InterruptedException , for instance when your code is part of a Runnable . In these situations, you must catch InterruptedException and restore the interrupted status by calling interrupt on the current thread, so that code higher up the call stack can see that an interrupt was issued, as demonstrated in Listing 5.10.

You can get much more sophisticated with interruption, but these two approaches should work in the vast majority of situations. But there is one thing you should not do with InterruptedException catch it and do nothing in response. This deprives code higher up on the call stack of the opportunity to act on the interruption, because the evidence that the thread was interrupted is lost. The only situation in which it is acceptable to swallow an interrupt is when you are extending Thread and therefore control all the code higher up on the call stack. Cancellation and interruption are covered in greater detail in Chapter 7.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 3 above)
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:740]: @veghlaci05 have you seen my comment above?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 9]
```diff
@@ -685,11 +672,16 @@ void open(CompactionInfo ci) throws TException {
             + "}, status {" + res.getState() + "}, reason {" + res.getErrorMessage() + "}");
       }
       lockId = res.getLockid();
-
-      heartbeatExecutor = Executors.newSingleThreadScheduledExecutor();
+      heartbeatExecutor = Executors.newSingleThreadScheduledExecutor(
+              CompactorUtil.createThreadFactory(
+                      "CompactionTxn Heartbeater - " + txnId, Thread.MIN_PRIORITY, true));
```
[the remark this answers is not in the document]
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:677]: nit: You could even avoid introducing a new utility method (which is used in only one place) and dependency to Guava via:
```
heartbeatExecutor = Executors.newSingleThreadScheduledExecutor((runnable) -> {
        Thread t = new Thread(runnable);
        t.setDaemon(true);
        t.setName("CompactionHeartbeater-" + txnId);
        t.setPriority(Thread.MIN_PRIORITY);
        return t;
      });
```

[the remark this answers is not in the document]
Comment (zabetak) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:677]: No strong feelings up to you :)

File: itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java [hunk 10]
```diff
@@ -170,6 +173,67 @@ public void tearDown() {
...
+    worker.setThreadId((int) worker.getId());
+    worker.setConf(conf);
+    worker.init(new AtomicBoolean(true));
+    FieldSetter.setField(worker, RemoteCompactorThread.class.getDeclaredField("msc"), mockedClient);
+
+    worker.run();
+
+    //Check if the heartbeating is properly terminated
+    Assert.assertTrue(Thread.getAllStackTraces().keySet()
+            .stream().noneMatch(k -> k.getName().contains("CompactionTxn Heartbeater")));
```
Comment (zabetak) [inline itests/hive-unit/src/test/java/org/apache/hadoop/hive/ql/txn/compactor/TestCompactor.java:234]: Maybe an assertion above that it really runs at some point could be useful.

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java [hunk 11]
```diff
@@ -657,24 +649,19 @@ private String getWorkerId() {
   /**
    * Keep track of the compaction's transaction and its operations.
    */
-  private class CompactionTxn implements AutoCloseable {
+  class CompactionTxn implements AutoCloseable {
```
Comment (zabetak) [inline ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:652]: Do we need to change visibility?

File: ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (veghlaci05) [reply ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/Worker.java:677]: I would keep it as-is, with Denys we think the utility method can be used elsewhere in the future, and the thread naming will also be more explicit this way.

**Review by zabetak — APPROVED** [review]
LGTM, thanks for your effort @veghlaci05 !
